### PR TITLE
Add the @ operator to the delayed objects

### PR DESCRIPTION
**Pull request by hmaarrfk**
not sure if this passes for python 2, but here goes.

- [x] Tests added / passed
- [x] Passes `flake8 dask`

### CODE REVIEW

File: dask/tests/test_delayed.py [hunk 1]
```diff
@@ -101,6 +101,14 @@ def test_operators():
...
     assert (a > 2).compute()
     assert (a ** 2).compute() == 100
+    if PY3:
+        class dummy:
+            def __matmul__(self, other):
+                return 4
+        c = delayed(dummy())
+        d = delayed(dummy())
+
+        assert (c @ d).compute() == 4
```
Comment (mrocklin) [inline dask/tests/test_delayed.py]: If this causes a SyntaxError in Python 2 (seems likely)  then we might wrap this in a string and then call `eval` like the following:

```python
In [1]: x = 1 

In [2]: eval('x + x')
Out[2]: 2
```

Comment (hmaarrfk) [reply dask/tests/test_delayed.py]: Hmm, you are right. Ok, let me try again
https://travis-ci.org/dask/dask/jobs/398449158#L1338

**Comment by hmaarrfk** [comment]
Is there a reason against iterating through `operator.__all__` and simply adding all of those functions to delayed objects?

```
import operator
for i in operator.__all__:
    print(getattr(operator, i))
```

**Comment by mrocklin** [comment]
I'll have to defer to @jcrist for that question.

**Comment by jcrist** [comment]
> Is there a reason against iterating through operator.__all__ and simply adding all of those functions to delayed objects?

We only want to support operators that can validly be delayed (e.g. `__add__` but not `__bool__`). There's also a lot of other things in `operator` that we'd have to filter out (e.g. `attrgetter`), so an explicit list is simpler and more robust.

PR looks good, thanks! Merging.

**Comment by jakirkham** [comment]
Thanks @hmaarrfk :)

**Comment by hmaarrfk** [comment]
No, thank you three, I was able to quite quickly (well after a bunch of refractoring) speedup my code quite s ignificantly using dask, 

I never use them, but is there a reason not to support operators like `+=`?

**Comment by jcrist** [comment]
> I never use them, but is there a reason not to support operators like +=?

These work automatically in python for things that don't mutate the object inplace. They work fine for delayed objects:

```python
In [1]: from dask import delayed

In [2]: a = delayed(1)

In [3]: a
Out[3]: Delayed('int-d1948b47-8998-4fbf-b52b-b5989688996c')

In [4]: a += 2

In [5]: a.compute()
Out[5]: 3
```

**Comment by hmaarrfk** [comment]
@jcrist clearly I never tried it. Python does a lot of magic sometimes ;)

File: dask/delayed.py [hunk 2]
```diff
@@ -520,6 +520,10 @@ def __call__(self, *args, **kwargs):
     Delayed._bind_operator(op)
 
 
+if PY3:
+    Delayed._bind_operator(operator.matmul)
```
Comment (jakirkham) [inline dask/delayed.py]: Guess we forgot about Python 3.4 (myself included). Put up a tweak in PR ( https://github.com/dask/dask/pull/3791 ), which should allow this to work on Python 3.4 as well.

**Comment by hmaarrfk** [comment]
I think the test needs to be tweeted as well in this case.

**Comment by jakirkham** [comment]
Good point. Thanks. Missed that.
